File: Services/SolarCalculator.cs
using System;

namespace SunTime.Services;

/// <summary>
/// Self-contained NOAA solar position algorithm.
/// Reference: https://gml.noaa.gov/grad/solcalc/solareqns.PDF
/// </summary>
public static class SolarCalculator
{
    public record SunData(
        DateTime Sunrise,
        DateTime Sunset,
        DateTime SolarNoon,
        double Altitude,
        double HourAngle);

    public static SunData Calculate(double latitude, double longitude, DateTime utcNow, TimeSpan utcOffset)
    {
        double jd = ToJulianDay(utcNow);
        double T = (jd - 2451545.0) / 36525.0;

        // Geometric Mean Longitude of Sun (deg)
        double L0 = Mod(280.46646 + T * (36000.76983 + 0.0003032 * T), 360);

        // Geometric Mean Anomaly of Sun (deg)
        double M = 357.52911 + T * (35999.05029 - 0.0001537 * T);

        // Eccentricity of Earth's orbit
        double e = 0.016708634 - T * (0.000042037 + 0.0000001267 * T);

        // Equation of Center (deg)
        double Mrad = Rad(M);
        double C = Math.Sin(Mrad) * (1.914602 - T * (0.004817 + 0.000014 * T))
                 + Math.Sin(2 * Mrad) * (0.019993 - 0.000101 * T)
                 + Math.Sin(3 * Mrad) * 0.000289;

        // Sun True & Apparent Longitude
        double sunTrueLong = L0 + C;
        double omega = 125.04 - 1934.136 * T;
        double sunAppLong = sunTrueLong - 0.00569 - 0.00478 * Math.Sin(Rad(omega));

        // Obliquity of Ecliptic
        double obliq0 = 23.0 + (26.0 + (21.448 - T * (46.815 + T * (0.00059 - T * 0.001813))) / 60.0) / 60.0;
        double obliq = obliq0 + 0.00256 * Math.Cos(Rad(omega));

        // Sun Declination (deg)
        double declination = Deg(Math.Asin(Math.Sin(Rad(obliq)) * Math.Sin(Rad(sunAppLong))));

        // Equation of Time (minutes)
        double y = Math.Pow(Math.Tan(Rad(obliq / 2)), 2);
        double eqTime = 4.0 * Deg(
            y * Math.Sin(2 * Rad(L0))
            - 2.0 * e * Math.Sin(Mrad)
            + 4.0 * e * y * Math.Sin(Mrad) * Math.Cos(2 * Rad(L0))
            - 0.5 * y * y * Math.Sin(4 * Rad(L0))
            - 1.25 * e * e * Math.Sin(2 * Mrad));

        // Solar Noon (minutes from midnight, local standard time)
        double solarNoonMin = 720.0 - 4.0 * longitude - eqTime + utcOffset.TotalMinutes;

        // Hour angle at sunrise/sunset (deg)
        double latRad = Rad(latitude);
        double decRad = Rad(declination);
        double cosHA = Math.Cos(Rad(90.833)) / (Math.Cos(latRad) * Math.Cos(decRad))
                     - Math.Tan(latRad) * Math.Tan(decRad);
        cosHA = Math.Clamp(cosHA, -1.0, 1.0); // polar day/night guard
        double ha = Deg(Math.Acos(cosHA));

        double sunriseMin = solarNoonMin - ha * 4.0;
        double sunsetMin  = solarNoonMin + ha * 4.0;

        var midnight = (utcNow + utcOffset).Date;
        var sunrise   = midnight.AddMinutes(sunriseMin);
        var sunset    = midnight.AddMinutes(sunsetMin);
        var solarNoon = midnight.AddMinutes(solarNoonMin);

        // Current altitude & hour angle
        double utcMinutes = utcNow.TimeOfDay.TotalMinutes;
        double tst = Mod(utcMinutes + eqTime + 4.0 * longitude, 1440);
        double hourAngle = tst / 4.0 - 180.0;

        double sinAlt = Math.Sin(latRad) * Math.Sin(decRad)
                      + Math.Cos(latRad) * Math.Cos(decRad) * Math.Cos(Rad(hourAngle));
        double altitude = Deg(Math.Asin(Math.Clamp(sinAlt, -1.0, 1.0)));

        return new SunData(sunrise, sunset, solarNoon, altitude, hourAngle);
    }

    // ── helpers ──────────────────────────────────────────────

    private static double Rad(double deg) => deg * Math.PI / 180.0;
    private static double Deg(double rad) => rad * 180.0 / Math.PI;
    private static double Mod(double a, double b) { double r = a % b; return r < 0 ? r + b : r; }

    private static double ToJulianDay(DateTime utc)
    {
        int y = utc.Year;
        int m = utc.Month;
        double d = utc.Day + utc.TimeOfDay.TotalDays;

        if (m <= 2) { y--; m += 12; }

        int A = y / 100;
        int B = 2 - A + A / 4;

        return (int)(365.25 * (y + 4716)) + (int)(30.6001 * (m + 1)) + d + B - 1524.5;
    }
}

File: apple-watch/SunTimeWatch Watch App/SolarCalculator.swift
import Foundation

enum SolarCalculator {
    static func calculate(latitude: Double, longitude: Double,
                          utcNow: Date, utcOffsetSeconds: Double) -> SunData {
        let jd = julianDay(utcNow)
        let T = (jd - 2451545.0) / 36525.0

        let L0 = fmod(280.46646 + T * (36000.76983 + 0.0003032 * T), 360.0)
        let M  = 357.52911 + T * (35999.05029 - 0.0001537 * T)
        let e  = 0.016708634 - T * (0.000042037 + 0.0000001267 * T)

        let Mrad = rad(M)
        let C = sin(Mrad) * (1.914602 - T * (0.004817 + 0.000014 * T))
              + sin(2 * Mrad) * (0.019993 - 0.000101 * T)
              + sin(3 * Mrad) * 0.000289

        let sunTrueLong = L0 + C
        let omega = 125.04 - 1934.136 * T
        let sunAppLong = sunTrueLong - 0.00569 - 0.00478 * sin(rad(omega))

        let obliq0 = 23.0 + (26.0 + (21.448 - T * (46.815 + T * (0.00059 - T * 0.001813))) / 60.0) / 60.0
        let obliq  = obliq0 + 0.00256 * cos(rad(omega))

        let declination = deg(asin(sin(rad(obliq)) * sin(rad(sunAppLong))))

        let y = pow(tan(rad(obliq / 2)), 2)
        let eqTime = 4.0 * deg(
            y * sin(2 * rad(L0))
            - 2.0 * e * sin(Mrad)
            + 4.0 * e * y * sin(Mrad) * cos(2 * rad(L0))
            - 0.5 * y * y * sin(4 * rad(L0))
            - 1.25 * e * e * sin(2 * Mrad))

        let offsetMin = utcOffsetSeconds / 60.0
        let solarNoonMin = 720.0 - 4.0 * longitude - eqTime + offsetMin

        let latRad = rad(latitude)
        let decRad = rad(declination)
        var cosHA = cos(rad(90.833)) / (cos(latRad) * cos(decRad)) - tan(latRad) * tan(decRad)
        cosHA = max(-1.0, min(1.0, cosHA))
        let ha = deg(acos(cosHA))

        let sunriseMin = solarNoonMin - ha * 4.0
        let sunsetMin  = solarNoonMin + ha * 4.0

        let cal = Calendar(identifier: .gregorian)
        let localNow = utcNow.addingTimeInterval(utcOffsetSeconds)
        var comps = cal.dateComponents([.year, .month, .day], from: localNow)
        comps.hour = 0; comps.minute = 0; comps.second = 0
        let midnight = cal.date(from: comps)!.addingTimeInterval(-utcOffsetSeconds)

        let sunrise   = midnight.addingTimeInterval(sunriseMin * 60)
        let sunset    = midnight.addingTimeInterval(sunsetMin  * 60)
        let solarNoon = midnight.addingTimeInterval(solarNoonMin * 60)

        let utcCal = Calendar(identifier: .gregorian)
        let utcComps = utcCal.dateComponents(in: TimeZone(identifier: "UTC")!, from: utcNow)
        let utcMinutes = Double((utcComps.hour ?? 0) * 60 + (utcComps.minute ?? 0))
                       + Double(utcComps.second ?? 0) / 60.0

        let tst = fmod(utcMinutes + eqTime + 4.0 * longitude + 1440.0, 1440.0)
        let hourAngle = tst / 4.0 - 180.0

        let sinAlt = sin(latRad) * sin(decRad)
                   + cos(latRad) * cos(decRad) * cos(rad(hourAngle))
        let altitude = deg(asin(max(-1.0, min(1.0, sinAlt))))

        return SunData(sunrise: sunrise, sunset: sunset, solarNoon: solarNoon,
                       altitude: altitude, hourAngle: hourAngle)
    }

    private static func rad(_ d: Double) -> Double { d * .pi / 180.0 }
    private static func deg(_ r: Double) -> Double { r * 180.0 / .pi }

    private static func julianDay(_ date: Date) -> Double {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "UTC")!
        let comps = cal.dateComponents([.year, .month, .day, .hour, .minute, .second], from: date)
        var y = comps.year!
        var m = comps.month!
        let d = Double(comps.day!) + (Double(comps.hour!) * 3600
              + Double(comps.minute!) * 60 + Double(comps.second!)) / 86400.0
        if m <= 2 { y -= 1; m += 12 }
        let A = y / 100
        let B = 2 - A + A / 4
        return Double(Int(365.25 * Double(y + 4716)))
             + Double(Int(30.6001 * Double(m + 1)))
             + d + Double(B) - 1524.5
    }
}
